test/run_test.py

```diff
@@ -934,76 +940,190 @@ def run_test_retries(
     sc_command = f"--sc={stepcurrent_key}"
+    step_current_file = (
+        REPO_ROOT / ".pytest_cache/v/cache/stepcurrent" / stepcurrent_key
+    )
+
+    def save_cache(cache):
+        def _save_keys(keys, file):
+            with open(f"{step_current_file}/{file}", "w") as f:
+                json.dump({k: v for k, v in cache.items() if k in keys}, f, indent=2)
+
+        _save_keys(["to_run", "prev_run", "already_ran"], "init")
+        _save_keys(["ended_at", "pytest_previous_status"], "active")
+
+    def load_cache():
+        with open(f"{step_current_file}/init") as f:
+            cache = json.load(f)
+        with open(f"{step_current_file}/active") as f:
+            return {**cache, **json.load(f)}
+
     while True:
+        start = time.time()
+        test_report_path = get_test_report_path()
         ret_code, _ = retry_shell(
-            command + [sc_command] + print_items,
+            command + [sc_command] + print_items + [f"--save-xml={test_report_path}"],
             test_directory,
             stdout=output,
             stderr=output,
             env=env,
             timeout=timeout,
             retries=0,  # no retries here, we do it ourselves, this is because it handles timeout exceptions well
         )
+        duration = time.time() - start
         ret_code = 0 if ret_code == 5 else ret_code
-        if ret_code == 0 and not sc_command.startswith("--rs="):
-            break  # Got to the end of the test suite successfully
         signal_name = f" ({SIGNALS_TO_NAMES_DICT[-ret_code]})" if ret_code < 0 else ""
         print_to_file(f"Got exit code {ret_code}{signal_name}")
-
-        # Read what just failed/ran
+        # Read what was previously run
         try:
-            with open(
-                REPO_ROOT / ".pytest_cache/v/cache/stepcurrent" / stepcurrent_key
-            ) as f:
-                current_failure = f.read()
+            cache = load_cache()
         except FileNotFoundError:
             print_to_file(
-                "No stepcurrent file found. Either pytest didn't get to run (e.g. import error)"
+                "No stepcurrent file found. Either pytest didn't run"
                 + " or file got deleted (contact dev infra)"
             )
-            break
+            return ret_code, False
+        if ret_code != 0 and "ended_at" not in cache:
+            print_to_file("Failed despite no tests running")
+            return 1, False
 
         env = try_set_cpp_stack_traces(env, command, set=False)
-        if ret_code != 0:
-            num_failures[current_failure] += 1
 
-        if ret_code == 0:
-            # Rerunning the previously failing test succeeded, so now we can
-            # skip it and move on
-            sc_command = f"--scs={stepcurrent_key}"
-            print_to_file(
-                "Test succeeeded in new process, continuing with the rest of the tests"
+        # Remove prev_run from to_run
+        cache["to_run"] = [
+            [x, y]
```

Comment:
nit: could you please replace x,y with slightly more descriptive variable names? (seems like they represent `test` and `status`)